### PR TITLE
64, 54 - Full Text Search and Fix Query By Date

### PR DESCRIPTION
**Pull request by ericrkuo**
https://github.com/ubclaunchpad/foodies/issues/64, https://github.com/ubclaunchpad/foodies/issues/54

When we call GET `promotions?searchQuery=beef $30`, we can now query using Postgres's full-text search to simulate a low-level tier google search engine.

This means that results are ordered by relevance to the search query, and we even have parts of the promotion name and description that are bolded when keywords match the search query. Postgres's full-text search would also find words based on lexemes, so for example, if our search query is `matches`, Postgres can find other words like `match, matching`.

Example response here. Then in FE, we can show the bold description and bold name.
image

Also fixed querying by expiration date. We now first convert all times to UTC and then compare them. For some reason, although `promotion.expirationDate` is already in UTC format, I still needed to include `at time zone UTC`. 

We can now filter promotions by expiration date accurately up to even the `seconds`.
https://github.com/ubclaunchpad/foodies/blob/c37de590a6144c4241fb5f80c9bb86c22315c842/backend/src/main/repository/PromotionRepository.ts#L55

Also updated all our sample data!

### CODE REVIEW

**Comment by yunishimura0716** [comment]
Great work! I'm just asking you about the interface of FullTextSearch. How do you get the value of `rank`? I'm sorry If I'm missing something

**Comment by ericrkuo** [comment]
> Great work! I'm just asking you about the interface of FullTextSearch. How do you get the value of `rank`? I'm sorry If I'm missing something

Thanks! No need to apologize, the `rank` comes from here.
https://github.com/ubclaunchpad/foodies/blob/c37de590a6144c4241fb5f80c9bb86c22315c842/backend/src/main/repository/PromotionRepository.ts#L79-L83

So we use `ts_rank_cd` which takes in two inputs, a `ts_vector` and `ts_query`. We optimized FullTextSearch by adding a new column to promotions called `tsvector` which we use as the first input to `ts_rank_cd`. And our `ts_query` comes from the user's search query they input. Then we leave the rest to Postgres to compute the rank!

**Comment by yunishimura0716** [comment]
> > Great work! I'm just asking you about the interface of FullTextSearch. How do you get the value of `rank`? I'm sorry If I'm missing something
> 
> Thanks! No need to apologize, the `rank` comes from here.
> https://github.com/ubclaunchpad/foodies/blob/c37de590a6144c4241fb5f80c9bb86c22315c842/backend/src/main/repository/PromotionRepository.ts#L79-L83
> 
> So we use `ts_rank_cd` which takes in two inputs, a `ts_vector` and `ts_query`. We optimized FullTextSearch by adding a new column to promotions called `tsvector` which we use as the first input to `ts_rank_cd`. And our `ts_query` comes from the user's search query they input. Then we leave the rest to Postgres to compute the rank!

I see, thanks. what does ts_vector represent?

**Comment by ericrkuo** [comment]
ts_vector is basically a document in a form thats optimized for full text search. So like it's broken down into something called lexemes.
https://www.postgresql.org/docs/10/datatype-textsearch.html This link does a really good explanation.

I configured in the BE that ts_vector contains lexemes for a promotions **name and description**, so the ts_vector represents two columns. It's really powerful, that's how postgres knows that words like `fry, fries, frying, fried` are all very similar.

https://github.com/ubclaunchpad/foodies/blob/c37de590a6144c4241fb5f80c9bb86c22315c842/backend/src/main/migration/1605475222018-FullTextSearch.ts#L15

**Comment by yunishimura0716** [comment]
> ts_vector is basically a document in a form thats optimized for full text search. So like it's broken down into something called lexemes.
> https://www.postgresql.org/docs/10/datatype-textsearch.html This link does a really good explanation.
> 
> I configured in the BE that ts_vector contains lexemes for a promotions **name and description**, so the ts_vector represents two columns. It's really powerful, that's how postgres knows that words like `fry, fries, frying, fried` are all very similar.
> 
> https://github.com/ubclaunchpad/foodies/blob/c37de590a6144c4241fb5f80c9bb86c22315c842/backend/src/main/migration/1605475222018-FullTextSearch.ts#L15

That's so cool! It's kind of technique of natural language processing! thank you so much for explaining!

**Review by yunishimura0716 — APPROVED** [review]

**Comment by ericrkuo** [comment]
Oh very interesting, I did not see that connection! I don't know too much about natural language processing, but I can see how ts_vector could represent some kind of pre-processing and cleaning of the data. Do you mind explaining or clarifying how this connects to NLP techniques based on your experience?

**Comment by yunishimura0716** [comment]
> Oh very interesting, I did not see that connection! I don't know too much about natural language processing, but I can see how ts_vector could represent some kind of pre-processing and cleaning of the data. Do you mind explaining or clarifying how this connects to NLP techniques based on your experience?

I'm not also expert of NLP and I just have used some library of NLP before. The thing ts_vector contains lexemes is similar to Morphological Analysis of NLP. Morphological Analysis is a kind of analysis in NLP, which makes Morpheme from text data based on the grammar and part of speech of the words. This analysis has been applied to many field of NLP like calculating the similarity between two sentences. I have used `Mecab`. If you're interested in, check it!

**Comment by ericrkuo** [comment]
Woahh i see very interesting thanks for the explanation. I'll check it out!

File: backend/src/main/repository/PromotionRepository.ts [hunk 1]
```diff
@@ -50,10 +50,14 @@ export class PromotionRepository extends Repository<Promotion> {
       });
     }
 
+    // see https://github.com/ubclaunchpad/foodies/issues/54
     if (promotionQuery?.expirationDate) {
-      queryBuilder.andWhere('promotion.expirationDate >= :date', {
-        date: promotionQuery.expirationDate,
-      });
+      queryBuilder.andWhere(
+        "promotion.expirationDate ::timestamptz at time zone 'UTC' >= :date ::timestamptz at time zone 'UTC'",
```
Comment (ld357) [inline backend/src/main/repository/PromotionRepository.ts:56]: Sorry if you explained this in the past and I'm forgetting but what's the purpose of this?

**Review by ld357 — COMMENTED** [review]
Nicely done!! This is so cool :) just had one question!

File: backend/src/main/repository/PromotionRepository.ts (hunk 1 above)
Comment (ericrkuo) [reply backend/src/main/repository/PromotionRepository.ts:56]: Oh this is if we want to filter promotions by expiration date, so for example if a user wants to find all promotions that are not expired after a given date. Does that answer your question?

Comment (ld357) [reply backend/src/main/repository/PromotionRepository.ts:56]: Oooh I see, yup that does!

**Review by ld357 — APPROVED** [review]

**Pull request merged**